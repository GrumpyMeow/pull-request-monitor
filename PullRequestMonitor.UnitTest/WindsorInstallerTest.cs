using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using Castle.Core.Internal;
using Castle.MicroKernel;
using Castle.MicroKernel.Handlers;
using Castle.Windsor;
using Castle.Windsor.Diagnostics;
using NUnit.Framework;

namespace PullRequestMonitor.UnitTest
{
    [TestFixture]
    public class WindsorInstallerTest : InstallerTest
    {
        public WindsorInstallerTest()
        {
            WindsorContainer = new WindsorContainer();
            WindsorContainer.Install(new WindsorInstaller());
        }

        protected override IWindsorContainer WindsorContainer { get; }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void CheckPotentiallyMisconfiguredComponents()
        {
            CheckPotentiallyMisconfiguredComponentsCore();
        }


        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void CheckNonPublicConstructors()
        {
            ApprovedNonPublicConstructorsCore();
        }
    }

    public abstract class InstallerTest
    {
        protected abstract IWindsorContainer WindsorContainer { get; }

        private static IEnumerable<IHandler> GetPotentiallyMisconfiguredComponents(IWindsorContainer container)
        {
            var host = container.Kernel.GetSubSystem(SubSystemConstants.DiagnosticsKey) as IDiagnosticsHost;
            var diagnostic = host.GetDiagnostic<IPotentiallyMisconfiguredComponentsDiagnostic>();
            var handlers = diagnostic.Inspect();
            return handlers;
        }

        private static IEnumerable<IHandler> GetAllComponents(IWindsorContainer container)
        {
            var host = container.Kernel.GetSubSystem(SubSystemConstants.DiagnosticsKey) as IDiagnosticsHost;
            var diagnostic = host.GetDiagnostic<IAllComponentsDiagnostic>();
            var handlers = diagnostic.Inspect();
            return handlers;
        }

        protected void CheckPotentiallyMisconfiguredComponentsCore()
        {
            var handlers = GetPotentiallyMisconfiguredComponents(WindsorContainer);
            var message = new StringBuilder();
            var inspector = new DependencyInspector(message);
            foreach (var handler in handlers)
            {
                ((IExposeDependencyInfo) handler).ObtainDependencyDetails(inspector);
            }
            Approvals.Verify(message.ToString());
        }

        protected void ApprovedNonPublicConstructorsCore()
        {
            var handlers = GetAllComponents(WindsorContainer);
            var message = new StringBuilder();
            foreach (var handler in handlers)
            {
                if (!handler.ComponentModel.HasClassServices)
                    continue;

                if (handler.ComponentModel.Constructors.IsNullOrEmpty())
                {
                    message.AppendFormat("Cannot construct service {0} due to missing public constructor", handler);
                    message.AppendLine();
                }
            }
            Approvals.Verify(message.ToString());
        }

        [Test]
        public void PrintContainer()
        {
            var handlers = GetAllComponents(WindsorContainer);
            Console.WriteLine("{0} components were registered.", handlers.Count());
            foreach (var handler in handlers)
            {
                Console.Write("Model: {0}", handler.ComponentModel.Implementation.Name);
                foreach (var service in handler.ComponentModel.Services)
                {
                    Console.Write(" / ");
                    Console.Write(service.Name);
                }
                Console.Write("\n");
            }
        }
    }
}