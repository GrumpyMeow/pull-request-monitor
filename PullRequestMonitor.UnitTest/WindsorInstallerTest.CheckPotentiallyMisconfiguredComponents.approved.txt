
'PullRequestMonitor.Model.NameRegexpRepositoryFilter' is waiting for the following dependencies:
- Parameter 'repoNamePattern' which was not provided. Did you forget to set the dependency?

'PullRequestMonitor.Model.PullRequest' is waiting for the following dependencies:
- Service 'Microsoft.TeamFoundation.SourceControl.WebApi.GitPullRequest' which was not registered.
- Parameter 'serverUri' which was not provided. Did you forget to set the dependency?
- Service 'PullRequestMonitor.Model.TfGitRepository' which was registered but is also waiting for dependencies.
'PullRequestMonitor.Model.TfGitRepository' is waiting for the following dependencies:
- Service 'Microsoft.TeamFoundation.SourceControl.WebApi.GitRepository' which was not registered.
- Service 'PullRequestMonitor.Model.TfProject' which was registered but is also waiting for dependencies.
'PullRequestMonitor.Model.TfProject' is waiting for the following dependencies:
- Service 'Microsoft.TeamFoundation.Core.WebApi.TeamProjectReference' which was not registered.
- Service 'PullRequestMonitor.Model.ITfsConnection' which was not registered.

'PullRequestMonitor.Model.TfProjectCollection' is waiting for the following dependencies:
- Parameter 'uri' which was not provided. Did you forget to set the dependency?
